Snap geneva updates

tonyespy opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit messages follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [x] Other... Please describe:

This PR includes multiple changes to the snap packaging for Geneva including:

 * update Consul, postgreSQL, and Vault to the latest versions used for Geneva
 * enable secure postgreSQL (issue #1943)
 * enable secure Redis
 * disable sys-mgmt-agent by default (see issue #2486)
 * include new device-virtual binary device profile
 * add Kuiper support
 * update release tags for non-edgex-go EdgeX services (e.g. device-virtual, support-rulesengine, ...)
 * fix configure hook logic to allow the database provider to be switched between mongo and redis (needed due to the removal of core-config-seed)
 * update the snap README file

**Note** - testing Kuiper just requires using 'snap set' to activate kuiper, then using the included kuiper-cli based on the upstream EdgeX instructions here:

https://github.com/emqx/kuiper/blob/master/docs/en_US/edgex/edgex_rule_engine_tutorial.md#kuiper-edgex-integration

## Does this PR introduce a breaking change?
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [x] No

## Are there any specific instructions or things that should be known prior to reviewing?

Full testing of security-proxy has been deferred, and will be undertaken as part of the snap kong update PR (pending).

andresrinivasan commented on snap/local/runtime-helpers/bin/security-secretstore-setup-launch.sh:
```diff
@@ -0,0 +1,23 @@
+
+security-secretstore-setup -confdir "$SNAP_DATA/config/security-secretstore-setup/res" --vaultInterval=10
+
+if [ -n "$REDIS5_PASSWORD_PATHNAME" ]; then
```
When would this be not true?

tonyespy replied:
Nice catch, since the env var is always set prior to this script being launched, the length of the string is always non-zero. Instead this should check if the file exists, and the security-secretstore-read command should also get moved into this if block. Here's the diff I'm testing now:
```

diff --git a/snap/local/runtime-helpers/bin/security-secretstore-setup-launch.sh b/snap/local/runtime-helpers/bin/security-secretstore-setup-launch.sh
index 234bd58..5cfc20a 100755
--- a/snap/local/runtime-helpers/bin/security-secretstore-setup-launch.sh
+++ b/snap/local/runtime-helpers/bin/security-secretstore-setup-launch.sh
@@ -2,7 +2,8 @@
 
 security-secretstore-setup -confdir "$SNAP_DATA/config/security-secretstore-setup/res" --vaultInterval=10
 
-if [ -n "$REDIS5_PASSWORD_PATHNAME" ]; then
+# if redis pw file doesn't exist, handle initialization & read it
+if [ ! -f "$REDIS5_PASSWORD_PATHNAME" ]; then
 
     CONFIG_FILE="$SNAP_DATA/config/security-secretstore-read/res/configuration.toml"
     # grep -q returns 0 if a match is found
@@ -18,6 +19,8 @@ if [ -n "$REDIS5_PASSWORD_PATHNAME" ]; then
 
         mv "$CONFIG_FILE.new" "$CONFIG_FILE"
     fi
+
+    # read the pw from secretstore & write to disk
+    security-secretstore-read -confdir "$SNAP_DATA/config/security-secretstore-read/res"
 fi
 
-security-secretstore-read -confdir "$SNAP_DATA/config/security-secretstore-read/res"

```

anonymouse64 replied:
also I have the same question, but for your new patch, under what conditions would security-secretstore-setup-launch.sh get run when this file doesn't exist?  Do we allow users to configure to use redis w/o a password

tonyespy replied:
At the moment, security-secretstore-setup-launch.sh is always run in order to launch security-secretstore-setup, which is a one-shot service. The file should not exist when this script is run.  We currently do not allow users to configure mongo or redis without password authentication.

tonyespy replied:
This has been addressed in my latest update

anonymouse64 commented on snap/local/runtime-helpers/bin/redis-launch.sh:
```diff
@@ -0,0 +1,16 @@
+exec "$SNAP/bin/redis-server" \
+     --requirepass "$REDIS5_PASSWORD" \
+     --dir "$SNAP_DATA/redis" \
+     --save 900 1 \
```
what do the --save args do?

anonymouse64 commented on snap/snapcraft.yaml:
```diff
@@ -155,13 +155,23 @@ apps:
+    environment:
+      VAULT_CONFIG: "$SNAP_DATA/config/security-secret-store/vault-config.hcl"
+      VAULT_ADDR: "https://localhost:8200"
+      VAULT_SKIP_VERIFY: "true"
```
why is this arg needed? If we need to skip verification when talking to a local vaul in the snap, I'd much rather the user have to provide an "skip-verify" arg than us always hard-coding it here

anonymouse64 replied:
Ping on this

anonymouse64 commented on snap/snapcraft.yaml:
```diff
@@ -155,13 +155,23 @@ apps:
+    adapter: none
+    command: bin/vault
+    environment:
+      VAULT_CONFIG: "$SNAP_DATA/config/security-secret-store/vault-config.hcl"
+      VAULT_ADDR: "https://localhost:8200"
```
if you wanted to be fancy, you could probably combine this with the yaml above with a yaml anchor. no preference from me, but if we need it for a third app I would prefer it.

anonymouse64 commented on snap/snapcraft.yaml:
```diff
@@ -327,7 +339,8 @@ apps:
   app-service-configurable:
     adapter: full
-    command: bin/app-service-configurable -confdir $SNAP_DATA/config/app-service-configurable/res -profile rules-engine --registry
+    # -s option disables the core-data version check; not needed in the snap since versions are known
```
:+1:

anonymouse64 commented on snap/snapcraft.yaml:
```diff
@@ -966,22 +1013,16 @@ parts:
-      if [ -f /build/go.mod ]; then
-        mv /build/go.mod /build/go.mod.bk
-      fi
+      patch -p1 < "$SNAPCRAFT_PROJECT_DIR/snap/local/patches/0001-Makefile-rm-golangci-lint-tool.patch"
```
could you add a comment about needing this patch?

anonymouse64 commented on snap/snapcraft.yaml:
```diff
@@ -1062,3 +1108,30 @@ parts:
+      bin/server: bin/kuiper-server
+    stage:
+      - -etc/mqtt_source.yaml
+    stage-packages:
+      - libzmq3-dev
```
this is already staged by the edgex-go part. snapcraft is smart enough to not care since the file contents will be the same AFAIK, but just noticing this

anonymouse64 replied:
also relatedly, why are we staging the dev package here, when we also stage `libzmq5` in app-service-config ? something smells wrong

anonymouse64 commented on snap/snapcraft.yaml:
```diff
@@ -400,19 +413,7 @@ apps:
-      # who is initially created as a role with postgres
-      - bin/drop
-    plugs: [home, removable-media, network]
+  # TODO: remove as part of secure postgres PR!
```
what should be removed as part of the secure postgres PR?

anonymouse64 replied:
(and is that this PR?)

tonyespy replied:
leftover comment, will remove when I push my update

anonymouse64 commented on snap/local/runtime-helpers/bin/security-secretstore-setup-launch.sh:
```diff
@@ -0,0 +1,23 @@
+if [ -n "$REDIS5_PASSWORD_PATHNAME" ]; then
+
+    CONFIG_FILE="$SNAP_DATA/config/security-secretstore-read/res/configuration.toml"
+    # grep -q returns 0 if a match is found
+    grep -q "/tmp/edgex" "$CONFIG_FILE"
+    RETVAL="$?"
+
+    if [ $RETVAL == 0 ]; then
```
why isn't this whole thing just 

```suggestion
    if grep -q "/tmp/edgex" "$CONFIG_FILE"; then
```

tonyespy replied:
Cause my shell fu isn't as strong as yours... thanks for the suggestion!

anonymouse64 commented on snap/local/runtime-helpers/bin/security-secretstore-setup-launch.sh:
```diff
@@ -0,0 +1,23 @@
+    RETVAL="$?"
+
+    if [ $RETVAL == 0 ]; then
+        sed -e "s@/tmp/edgex/secrets/ca/ca.pem@$SNAP_DATA/pki/EdgeXFoundryCA/EdgeXFoundryCA.pem@" \
+            -e "s@/tmp/edgex/secrets/edgex-redis/secrets-token.json@$SNAP_DATA/secrets/edgex-redis/secrets-token.json@" \
+            -e "s@/tmp/edgex/secrets/edgex-redis/redis5-password@$REDIS5_PASSWORD_PATHNAME@" \
+            -e "s@edgex-vault@localhost@" \
```
why are we performing this sed expansion at runtime? I'd much rather do this at build-time or if really needed install-hook time, not at app service runtime

anonymouse64 commented on snap/hooks/install:
```diff
@@ -10,14 +10,15 @@ SNAP_CURRENT=${SNAP/%$SNAP_REVISION/current}
 # note that app-service-configurable is handled separately
 mkdir -p "$SNAP_DATA/config"
-for service in edgex-mongo security-file-token-provider security-proxy-setup security-secrets-setup security-secretstore-setup core-command core-data core-metadata support-logging support-notifications support-scheduler sys-mgmt-agent device-virtual; do
+for service in edgex-mongo security-file-token-provider security-proxy-setup security-secrets-setup security-secretstore-setup core-command core-data core-metadata support-logging support-notifications support-scheduler sys-mgmt-agent device-virtual security-secretstore-read; do
```
is this a new service? I don't see the service defined in the snapcraft.yaml ?

tonyespy replied:
Yes, this is a new service in edgex-go which reads the redis password from vault, and writes it to the redis password file. It's included in the edgex-go part's build-override logic (see the first for loop which iterates over the all the services to copy the binaries and configuration files).

anonymouse64 replied:
I guess I was more wondering why it's not a proper app/service in the yaml?

tonyespy replied:
Because it's only ever called once from the wrapper script to read the Redis password from Vault, and write it to disk. I treated this the same as security-file-token-provider which is launched from security-secret-store-setup. Ideally, Redis should just grow support for reading secrets directly from Vault, but that wasn't scoped for this release.

anonymouse64 replied:
ok, thanks for the explanation

tonyespy commented on snap/hooks/configure:
```diff
@@ -126,75 +144,134 @@ done
+                if [ "$exists" = "redisdb" ]; then
+
+                    curl -s --request PUT --data "mongodb" \
+                        http://localhost:8500/v1/kv/edgex/core/1.0/edgex-$svc/Databases/Primary/Type
```
Is this what you were thinking? I'm not sure I have the cycles to create a single function that handles all of the logic for each db type?

```
exists=`curl -s http://localhost:8500/v1/kv/edgex/core/1.0/edgex-$svc/Databases/Primary/Type?raw`
if [ "$exists" = "redisdb" ]; then
    update_consul_db_type "$svc" "mongodb" "27017"
    updated="true"
fi
```

anonymouse64 replied:
yes

tonyespy commented on snap/hooks/install:
```diff
@@ -156,33 +162,48 @@ snapctl start "$SNAP_NAME.postgres"
 # stop postgres again in case the security services should be turned off
 snapctl stop "$SNAP_NAME.postgres"
 
+# modify postgres authentication config to use 'md5' (password)
```
I'm loathe to change this, as it doesn't make sense to configure postgres to require password authentication before creating the user that owns the credentials...

anonymouse64 replied:
okay, fair enough

tonyespy commented:
I just pushed an update that addresses most of the comments. The only thing I didn't update was the install script logic which enables postgresql password auth, as it doesn't make sense to enable password auth before the user (with password) is added.

I also added one more commit which enables configure hook support for the SMA, as it's now disabled by default.

tonyespy commented:
And I just snuck in one additional commit which updates mongo to the latest stable patch release (4.2.6).

ernestojeda commented:
recheck

lenny-intel commented on snap/hooks/configure:
```diff
@@ -52,6 +61,28 @@ handle_svc () {
+}
+
+update_consul_db_type () {
+    curl -s --request PUT --data "$2" \
```
Why Consul and not env overrides? 

Remember that env overrides are record of truth now, not consul.

tonyespy replied:
Well... back when core-config-seed still was around, this code modified the configuration.toml files in **$SNAP_DATA**, stopped the affected services, ran core-config-seed w/the overwrite flag, and then restarted the services. Doing this on a per-service basis seemed a bit too complex. Yes, it would be possible to do this via env overrides, but that would require an additional wrapper script (see bin/security-secret-store-env-var.sh which is used to handle vault being enabled/disabled). When Ian and I first discussed how to do this, he suggested just modifying the values directly in Consul, so I just went with that approach.

My gut tells me that modifying the dbtype (which needs to be done *before* anything's been written to the db) is going to be a fairly uncommon operation. The only reason I've kept it around is that we still haven't officially dropped mongo yet.

If you feel strongly about this, I'll go ahead and create the wrapper scripts.

lenny-intel replied:
Do you have a mode where Consul isn't used? If not, then I am fine with this approach.

anonymouse64 replied:
Since we don't yet support running without consul, I'd say leave it like this. Setting env vars for other services is also awkward because you need to pass state to those other services from here via either snap config or a file or something so it's not quite as simple as just setting the env var.

lenny-intel commented on snap/snapcraft.yaml:
```diff
@@ -426,6 +425,27 @@ apps:
+      # KUIPER_DEBUG: "true"
+      KUIPER_CONSOLE_LOG: "true"
+      KUIPER_REST_PORT: 48075
+      EDGEX_SERVER: edgex-app-service-configurable-rules
```
This should be `localhost` not the docker host name.

lenny-intel commented on snap/snapcraft.yaml:
```diff
@@ -526,43 +547,72 @@ parts:
   # eventually we would like these changes to be upstreamed but until then
   # they live here. see also https://github.com/edgexfoundry/edgex-go/issues/1109
-  go:
+  go112:
```
why still need go112?

tonyespy replied:
Because the snap builds consul and vault from source, and vault 1.3.1 can only be built with go 1.12.

lenny-intel commented on snap/snapcraft.yaml:
```diff
@@ -426,6 +425,27 @@ apps:
+      KUIPER_REST_PORT: 48075
+      EDGEX_SERVER: edgex-app-service-configurable-rules
+      EDGEX_SERVICE_SERVER: http://localhost:48080
+      EDGEX_TOPIC: events
```
Also don't see where you set `PublishTopic` on app-service-rules to be `events`. By default it is blank for the old rules engine.

lenny-intel replied:
FYI, this is the override we have on app-service-rules for Kuiper.       
```
Binding_PublishTopic: events
```

tonyespy commented on snap/README.md:
```diff
@@ -115,19 +122,19 @@ $ journalctl -u snap.edgexfoundry.consul
+ * security-secrets-setup
+ * security-secretstore-setup
+ * security-token-file-provider
+ * When
```
Whoops, nice catch.

tonyespy commented on snap/README.md:
```diff
@@ -115,19 +122,19 @@ $ journalctl -u snap.edgexfoundry.consul
-All services are currently bundled in the singular service, `security-services` (see issue [#485](https://github.com/edgexfoundry/edgex-go/issues/485) for more details on why). 
-
-When security is enabled, Consul is secured using Vault for secret management, and Kong is used as an HTTPS proxy for all the services. The HTTPS keys for Kong and Vault are placed in `$SNAP_DATA/vault/pki`. Kong needs a database to manage itself, and can use either Postgres or Cassandra. Because Postgres cannot run inside of a snap due to issues running as root (currently all snap services must  ...
+security-proxy-setupsecurity is enabled, Consul is secured using Vault for secret management, and Kong is used as an HTTPS proxy for all the services.
```
Looks like this whole section got mangled during a rebase. Fixed, and will push shortly.

lenny-intel commented on snap/snapcraft.yaml:
```diff
@@ -329,8 +341,9 @@ apps:
     command: bin/app-service-configurable -confdir $SNAP_DATA/config/app-service-configurable/res -profile rules-engine --registry
     daemon: simple
+    environment:
+      Binding_PublishTopic: events
```
FYI, I started to have concern that this would break the ability to switch to old rules engine that subscribes to blank topic, but I think the blank topic is like a wild card topic for ZMQ. Thoughts?

lenny-intel replied:
I just confirmed this is correct:
```
A consequence of this prefix matching behaviour is that you can receive all published messages by subscribing with an empty topic string.
```

tonyespy replied:
OK, thanks!

anonymouse64 commented on snap/hooks/configure:
```diff
@@ -52,6 +61,28 @@ handle_svc () {
+    # changes need to be progpogated into relevant services, so
+    # restart the services with updated configuration if they were
+    # already running
+    for svc in ${DB_SERVICES[*]}; do
```
I have a slight preference for helper functions to take arguments rather than use global variables but this is fine if you would rather change this later

tonyespy replied:
I just followed your lead (e.g. `ALL_SERVICES`) on this one. Let's leave this one alone.

anonymouse64 replied:
ah right I had missed that this variable was just globally assigned once, I thought it was mutated in the other functions, nvm me this is fine

anonymouse64 commented on snap/hooks/configure:
```diff
@@ -126,75 +170,103 @@ done
+
+                exists=`curl -s http://localhost:8500/v1/kv/edgex/core/1.0/edgex-$svc/Databases/Primary/Type?raw`
+                if [ "$exists" = "true" ]; then
+                    update_consul_db_type "$svc" "redisdb" "6379"
```
Thanks for this

anonymouse64 commented on snap/hooks/install:
```diff
@@ -156,33 +162,48 @@ snapctl start "$SNAP_NAME.postgres"
+drop sed -i -e "s@trust@md5@g" "$SNAP_DATA/postgresql/10/main/pg_hba.conf"
+
 # finally, disable and turn off non-default services
 # by default, we want the support-*, device-*, and mongo services
```
Would still be nice to see the sma here

anonymouse64 commented on snap/local/runtime-helpers/bin/security-secretstore-setup-launch.sh:
```diff
@@ -0,0 +1,11 @@
+security-secretstore-setup -confdir "$SNAP_DATA/config/security-secretstore-setup/res" --vaultInterval=10
+
+# if redis pw file doesn't exist, handle initialization & read it
+if [ ! -f "$REDIS5_PASSWORD_PATHNAME" ]; then
```
I'm still not clear on when this script is called when there is not a password defined?

tonyespy replied:
This script gets called once, on first boot when the redis password file doesn't exist. This is a workaround to the fact that redis can't read the password directly from Vault.

anonymouse64 commented on snap/snapcraft.yaml:
```diff
@@ -676,18 +720,36 @@ parts:
+              # since security-secretstore-read doesn't support config env overrides, edit the
+              # configuration file at build time
+              CONFIG_FILE="./cmd/$service/res/configuration.toml"
+              if grep -q "/tmp/edgex" "$CONFIG_FILE"; then
```
Thanks for moving this here

tonyespy replied:
What was the ping for?

anonymouse64 replied:
you missed a comment from my previous review about why we are hard-coding `VAULT_SKIP_VERIFY: "true"`

tonyespy commented on snap/local/patches/0001-Makefile-rm-golangci-lint-tool.patch:
```diff
@@ -0,0 +1,28 @@
+Date: Fri, 24 Apr 2020 08:33:04 -0400
+Subject: [PATCH] Makefile: rm golangci-lint tool
+
+Remopve golangci-lint from EXTERNAL_TOOLS as
```
https://github.com/hashicorp/vault/issues/8927

tonyespy commented:
All comments should be addressed, and I fixed a few other issues that cropped up during testing.

anonymouse64 commented on snap/hooks/configure:
```diff
@@ -126,75 +174,103 @@ done
+            for svc in ${DB_SERVICES[*]}; do
+                updated="false"
+
+                exists=`curl -s http://localhost:8500/v1/kv/edgex/core/1.0/edgex-$svc/Databases/Primary/Type?raw`
```
this logic doesn't quite work right, when I run this curl command locally I see this as returning the value of the `Type` key, which never ends up being "true" as you check below:

```
$ curl -s http://localhost:8500/v1/kv/edgex/core/1.0/edgex-core-data/Databases/Primary/Type?raw; echo
mongodb
```

anonymouse64 commented on snap/local/runtime-helpers/bin/security-secretstore-setup-launch.sh:
```diff
@@ -0,0 +1,11 @@
+security-secretstore-setup -confdir "$SNAP_DATA/config/security-secretstore-setup/res" --vaultInterval=10
+
+# if redis pw file doesn't exist, handle initialization & read it
+if [ ! -f "$REDIS5_PASSWORD_PATHNAME" ]; then
```
Sorry to be a bother about this, but I'm still rather confused, since AFAICT, the thing that will generate the password file is `security-secretstore-setup` which we literally call right before this check in this script, so the only situation that this condition would be false is if security-secretstore-setup is configured to not run generate a redis password? Is that a supported configuration for the snap to configure security-secretstore-setup to _not_ use a password for redis? That also won't be the default configuration since it seems to generate this password on initial install, so not only would the user need to configure `security-secretstore-setup` to not generate a password, but they would also need to _delete_ the existing password too. Seems like a lot of work and unlikely to ever be done by a user.

All that is to say, I think we could probably get away with not having this check.

tonyespy replied:
In a perfect world, security-secretstore-setup would redis would read it's password directly from vault, but this isn't supported. In a less perfect world, `security-secretstore-setup` would read the password from vault and write it out to disk, but this isn't supported either. In the docker deployment, `security-secretstore-read` is called from the [entrypoint.sh](https://github.com/edgexfoundry/edgex-go/blob/master/cmd/security-secretstore-setup/entrypoint.sh#L54), which is why this wrapper script exists in the first place. So, when this script is first called, the password file will **not** exist, hence the check...

anonymouse64 commented on snap/hooks/install:
```diff
@@ -198,14 +230,3 @@ for svc in support-notifications support-scheduler support-logging app-service-c
-# openresty
-# see https://github.com/edgexfoundry/blackbox-testing/issues/185 for more 
-# details
-if [ "$SNAP_ARCH" == "arm64" ]; then
-    snapctl set security-proxy=off
-    snapctl stop --disable "$SNAP_NAME.kong-daemon"
-    snapctl stop --disable "$SNAP_NAME.security-proxy-setup"
-fi
```
hmm I just tried my kong snap on my rpi on arm64 and it didn't seem to work, so I think we need to leave this in for users to be able to use the snap on arm64 at all

anonymouse64 commented on snap/snapcraft.yaml:
```diff
@@ -155,13 +155,23 @@ apps:
+    environment:
+      VAULT_CONFIG: "$SNAP_DATA/config/security-secret-store/vault-config.hcl"
+      VAULT_ADDR: "https://localhost:8200"
+      VAULT_SKIP_VERIFY: "true"
```
just fyi I still see this here

anonymouse64 commented on snap/snapcraft.yaml:
```diff
@@ -966,22 +1035,16 @@ parts:
-      if [ -f /build/go.mod ]; then
-        mv /build/go.mod /build/go.mod.bk
-      fi
+      patch -p1 < "$SNAPCRAFT_PROJECT_DIR/snap/local/patches/0001-Makefile-rm-golangci-lint-tool.patch"
```
I would still like to see a comment about why we need this patch and a link to the upstream issue here

tonyespy commented:
> Really really close, but...
> 
>     * mongodb -> redis logic seems to be broken
> 
>     * I don't think we're ready to re-enable arm64 yet
> 
>     * I would still like to see the check on the redis password eliminated since it doesn't seem to be needed as it will always exist
> 
>     * the vault skip_verify env var is still set
> 
>     * still no comment where we apply the patch to vault to build it without the CI failure

@anonymouse64 All of your comments have been addressed, and just re-pushed.
